realpath'ing Repo path -vs- ok_file_under_git

Originally ran into within https://github.com/datalad/datalad/pull/2898 ([failed travis run](https://travis-ci.org/datalad/datalad/jobs/442291459)), not sure how it didn't float up before (at least for me):  
in [GitRepo._flyweight_id_from_args](https://github.com/datalad/datalad/blob/master/datalad/support/gitrepo.py#L576) we realpath'ing every path so any repository's path is the realpath (although we do have `.realpath` for that purpose).  But in other spots (like `ok_file_under_git`) we seems to not do that at causes incorrect assessment.  Here are examples in both DataLad proper and the -revolution one:

```shell
$> python -c 'from datalad.support.gitrepo import GitRepo; p="/var/tmp/sym link/datalad_temp_tree_wrappedVlKRSJ/ds-clone"; print(GitRepo(p)); from datalad.tests.utils import ok_file_under_git; ok_file_under_git(p + "/in-annex", annexed=True)' 
<GitRepo path=/tmp/datalad_temp_tree_wrappedVlKRSJ/ds-clone (<class 'datalad.support.gitrepo.GitRepo'>)>
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-master/datalad/tests/utils.py", line 226, in ok_file_under_git
    assert_in(file_repo_path, repo.get_indexed_files())  # file is known to Git
  File "/usr/lib/python2.7/unittest/case.py", line 804, in assertIn
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: '../../../var/tmp/sym link/datalad_temp_tree_wrappedVlKRSJ/ds-clone/in-annex' not found in [u'.gitattributes', u'.datalad/.gitattributes', u'in-annex', u'.datalad/config', u'in-git']
(dev) 1 14949 ->1.....................................:Mon 29 Oct 2018 08:35:53 PM EDT:.
(git)hopa:~datalad/datalad-revolution[master]git
$> python -c 'from datalad_revolution.gitrepo import RevolutionGitRepo as GitRepo; p="/var/tmp/sym link/datalad_temp_tree_wrappedVlKRSJ/ds-clone"; print(GitRepo(p)); from datalad.tests.utils import ok_file_under_git; ok_file_under_git(p + "/in-annex", annexed=True)'             
<GitRepo path=/tmp/datalad_temp_tree_wrappedVlKRSJ/ds-clone (<class 'datalad_revolution.gitrepo.RevolutionGitRepo'>)>
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-master/datalad/tests/utils.py", line 226, in ok_file_under_git
    assert_in(file_repo_path, repo.get_indexed_files())  # file is known to Git
  File "/usr/lib/python2.7/unittest/case.py", line 804, in assertIn
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: '../../../var/tmp/sym link/datalad_temp_tree_wrappedVlKRSJ/ds-clone/in-annex' not found in [u'.gitattributes', u'.datalad/.gitattributes', u'in-annex', u'.datalad/config', u'in-git']
```

Quick solution of ok_file_under_git adjustment would be ok, but I am not sure if that would be the "correct" one...

--- a/datalad/tests/test_tests_utils.py
+++ b/datalad/tests/test_tests_utils.py
@@ -34,15 +34,17 @@ from nose import SkipTest
 from ..utils import getpwd, chpwd
 
 from . import utils
-from .utils import eq_, ok_, assert_false, ok_startswith, nok_startswith, \
-    with_tempfile, with_testrepos, with_tree, \
-    rmtemp, \
-    OBSCURE_PREFIX, OBSCURE_FILENAMES,\
-    get_most_obscure_supported_name, \
-    swallow_outputs, swallow_logs, \
-    on_windows, assert_raises, assert_cwd_unchanged, serve_path_via_http, \
-    ok_symlink, assert_true, ok_good_symlink, ok_broken_symlink
-
+from .utils import (
+    eq_, ok_, assert_false, ok_startswith, nok_startswith,
+    with_tempfile, with_testrepos, with_tree,
+    rmtemp,
+    OBSCURE_PREFIX, OBSCURE_FILENAMES,
+    get_most_obscure_supported_name,
+    swallow_outputs, swallow_logs,
+    on_windows, assert_raises, assert_cwd_unchanged, serve_path_via_http,
+    ok_symlink, assert_true, ok_good_symlink, ok_broken_symlink,
+    ok_file_under_git
+)
 from .utils import ok_generator
 from .utils import assert_dict_equal
 from .utils import assert_str_equal
@@ -50,6 +52,7 @@ from .utils import assert_re_in
 from .utils import local_testrepo_flavors
 from .utils import skip_if_no_network
 from .utils import skip_if_no_module
+from .utils import skip_if_on_windows
 from .utils import run_under_dir
 from .utils import skip_if
 from .utils import ok_file_has_content
@@ -60,6 +63,8 @@ from .utils import probe_known_failure
 from .utils import patch_config
 from .utils import ignore_nose_capturing_stdout
 
+from ..support.gitrepo import GitRepo
+from ..support import path as op
 #
 # Test with_tempfile, especially nested invocations
 #
@@ -634,3 +639,23 @@ def test_ignore_nose_capturing_stdout():
                              'StringIO and fileno')
     with assert_raises(SkipTest):
         ignore_nose_capturing_stdout(raise_exc)()
+
+
+@skip_if_on_windows  # no symlinks. may be skip if not hasattr(os, "symlink")
+@with_tree(tree={'ingit': '', 'staged': 'staged', 'notingit': ''})
+def test_ok_file_under_git_symlinks(path):
+    # Test that works correctly under symlinked path
+    orepo = GitRepo(path)
+    orepo.add('ingit')
+    orepo.commit('msg')
+    orepo.add('staged')
+    lpath = path + "-symlink"  # will also be removed AFAIK by our tempfile handling
+    os.symlink(path, lpath)
+    ok_symlink(lpath)
+    ok_file_under_git(op.join(path, 'ingit'))
+    ok_file_under_git(op.join(lpath, 'ingit'))
+    ok_file_under_git(op.join(lpath, 'staged'))
+    with assert_raises(AssertionError):
+        ok_file_under_git(op.join(lpath, 'notingit'))
+    with assert_raises(AssertionError):
+        ok_file_under_git(op.join(lpath, 'nonexisting'))

--- a/datalad/tests/utils.py
+++ b/datalad/tests/utils.py
@@ -280,7 +280,7 @@ def _prep_file_under_git(path, filename):
 
     # path to the file within the repository
     # repo.path is a "realpath" so to get relpath working correctly
-    # we need to relpath our path as well
+    # we need to realpath our path as well
     path = op.realpath(path)  # intentional realpath to match GitRepo behavior
     file_repo_dir = op.relpath(path, repo.path)
     file_repo_path = filename if file_repo_dir == curdir else opj(file_repo_dir, filename)

--- a/datalad/tests/utils.py
+++ b/datalad/tests/utils.py
@@ -279,7 +279,10 @@ def _prep_file_under_git(path, filename):
             raise
 
     # path to the file within the repository
-    file_repo_dir = os.path.relpath(path, repo.path)
+    # repo.path is a "realpath" so to get relpath working correctly
+    # we need to relpath our path as well
+    path = op.realpath(path)  # intentional realpath to match GitRepo behavior
+    file_repo_dir = op.relpath(path, repo.path)
     file_repo_path = filename if file_repo_dir == curdir else opj(file_repo_dir, filename)
     return annex, file_repo_path, filename, path, repo
 

